File: Advent of Code 2022/day5-p1.py
import re

problem_file = "input/day5.txt"

with open(problem_file) as file:
    input_lines = file.readlines()

empty_line_index = input_lines.index("\n")
stacks_list = input_lines[:empty_line_index - 1]
rearr_list = input_lines[empty_line_index + 1:]

stacks = []
i = 1
while True:
    stacks.append([])
    for line in reversed(stacks_list):
        elem = line[i]
        if elem != " ":
            stacks[-1].append(elem)
    i += 4
    if i >= len(stacks_list[0]):
        break

# stacks = [[elem for i in range(len(stacks_list)) if (elem := stacks_list[i][row]) != " "] for row in range(1, len(stacks_list[0]), 4)]

regex = re.compile(r".*?(\d+).*?(\d+).*?(\d+)")
for line in rearr_list:
    count, from_, to = map(lambda x: int(x), regex.match(line).group(1, 2, 3))
    stacks[to-1] += reversed(stacks[from_-1][-count:])
    del stacks[from_-1][-count:]

print(stacks)
print("".join(l[-1] for l in stacks))

File: Advent of Code 2022/day5-p1.exs
problem_file = "input/day5.txt"

[stacks_string, rearrangements_string] = File.read!(problem_file)
    |> String.split("\n\n", trim: true)

{stacks_list, [stack_indices]} = stacks_string
    |> String.split("\n", trim: true)
    |> Enum.split(-1)

stacks_list = stacks_list
    |> Enum.map(&String.to_charlist/1)
    |> Enum.map(&Enum.drop(&1, 1))
    |> Enum.map(&Enum.chunk_every(&1, 1, 4, :discard))

stacks = stacks_list |> Enum.zip_with(fn l -> Enum.reject(l, &(&1 == ' ')) end)
stacks = for {val, i} <- Enum.with_index(stacks), into: %{}, do: {i + 1, val}

rearrangements_list = rearrangements_string |> String.split("\n", trim: true)

rearrangements_list = rearrangements_list
    |> Enum.map(fn (elem) ->
        Regex.run(~r".+?(\d+).+?(\d+).+?(\d+)", elem, capture: :all_but_first) |> Enum.map(&String.to_integer/1)
    end)

stacks = Enum.reduce(rearrangements_list, stacks, fn ([count, from, to], stacks_) ->
    stacks_
        |> Map.put(to, Enum.reverse(Enum.take(stacks_[from], count)) ++ stacks_[to])
        |> Map.put(from, Enum.drop(stacks_[from], count))
    end)

IO.inspect(stacks)

solution = Enum.reduce(Map.values(stacks), [], &([List.first(&1, "") | &2])) |> Enum.reverse

IO.puts(solution)
